mempool/src/shared_mempool.rs

```diff
@@ -449,3 +435,42 @@ where
 
     runtime
 }
+
+struct Semaphore {
+    lock: Mutex<usize>,
+    cvar: Condvar,
+}
+
+impl Semaphore {
+    #[allow(clippy::mutex_atomic)]
+    fn new(count: usize) -> Self {
+        Self {
+            lock: Mutex::new(count),
+            cvar: Condvar::new(),
+        }
+    }
+
+    #[allow(clippy::mutex_atomic)]
+    fn acquire(&self) {
```

Comment:
Personally, I would prefer a more RAII approach. For example, `Semaphore::acquire` returns a `Permit` which functions like a `std::sync::MutexGuard`. When you drop the `Permit` it just calls `Semaphore::release`. This way it is more difficult to forget to call `Semaphore::release`. In other words,

```rust
struct Permit(Semaphore);

impl Drop for Permit {
    fn drop(&mut self) {
        self.0.release()
    }
}

struct Semaphore(Arc<Inner>);

struct Inner {
    lock: Mutex<usize>,
    cvar: Condvar,
}

impl Semaphore {
    fn acquire(&self) -> Permit;
}
```